src/session/mod.rs

```diff
@@ -367,6 +367,26 @@ impl Session {
         }
     }
 
+    pub(crate) fn cancel_file_download(&self, file_id: i32) {
+        let mut downloading_files = self.imp().downloading_files.borrow_mut();
+
+        if let Some(senders) = downloading_files.remove(&file_id) {
```

Comment:
I think that you should not remove senders here. You will receive a new file update with `is_downloading_active: false` when the file is canceled. So, here you just need to call `tdlib::cancel_download_file`, and you should remove the senders in the `handle_file_update` phase.

Reply:
Maybe we don't need to delete file here too, because telegram for android don't do this
But I don't know why

Reply:
> Maybe we don't need to delete file here too, because telegram for android don't do this But I don't know why

Yeah probably. I guess it's for cache reasons. We can implement the automatic cache deletion in the future anyway, so it won't be a problem.

Reply:
Ok also cancel_download_file returns result, what I need to do with it?

Reply:
I think you just need to show a warning.